whittle/models/gpt/blocks/causal_self_attention.py

```diff
@@ -45,86 +45,66 @@ def __init__(self, config: Config, block_idx: int) -> None:
-            end_queries = self.config.n_head * self.config.head_size
-            qkv_indices.extend(
-                [i for i in range(end_queries, end_queries + self.sub_network_head_size)]
-            )
-            end_keys = end_queries + self.config.head_size
-            qkv_indices.extend(
-                [i for i in range(end_keys, end_keys + self.sub_network_head_size)]
-            )
+        indices = []
+        hs = self.config.head_size
+        snhs = self.sub_network_head_size
+        n_head = self.config.n_head
+        n_qg = self.config.n_query_groups
+        snh = self.sub_network_n_head
+
+        heads_per_group = n_head // n_qg
+
+        if n_head == n_qg:
+            for h in range(snh):
+                base = 3 * h * hs
+                indices.append(torch.arange(base, base + snhs))  # q
+                indices.append(torch.arange(base + hs, base + hs + snhs))  # k
+                indices.append(torch.arange(base + 2 * hs, base + 2 * hs + snhs))  # v
+
+        elif n_qg == 1:
+            for h in range(snh):
+                start = h * hs
+                indices.append(torch.arange(start, start + snhs))  # q
+            end_queries = n_head * hs
+            indices.append(torch.arange(end_queries, end_queries + snhs))  # k
+            indices.append(torch.arange(end_queries + hs, end_queries + hs + snhs))  # v
+
         else:
             for g in range(self.sub_network_query_groups):
-                start_q = g * (heads_per_group + 2) * self.config.head_size
+                base_q = g * (heads_per_group + 2) * hs
                 for h in range(self.sub_network_q_per_kv):
-                    qkv_indices.extend(
-                        [
-                            i
-                            for i in range(
-                                start_q + h * self.config.head_size,
-                                start_q
-                                + h * self.config.head_size
-                                + self.sub_network_head_size,
-                            )
-                        ]
-                    )
-                start_k = start_q + heads_per_group * self.config.head_size
-                qkv_indices.extend(
-                    [i for i in range(start_k, start_k + self.sub_network_head_size)]
-                )
-                start_v = start_k + self.config.head_size
-                qkv_indices.extend(
-                    [i for i in range(start_v, start_v + self.sub_network_head_size)]
-                )
-        return qkv_indices
+                    indices.append(
+                        torch.arange(base_q + h * hs, base_q + h * hs + snhs)
+                    )  # q
+                base_k = base_q + heads_per_group * hs
+                indices.append(torch.arange(base_k, base_k + snhs))  # k
+                indices.append(torch.arange(base_k + hs, base_k + hs + snhs))  # v
+
+        return torch.cat(indices)
```

Comment:
To be exactly equivalent to the previous implementation, the tensor has to be cast into a list.  The same applies for the function below as well.

Other than that, LGTM!